### PR TITLE
Amend FramedVoronoiGrid issue #1450 #1454 follow-up

### PR DESCRIPTION
**Pull request by sebastien-lenard**
### Description
Amends FramedVoronoiGrid with general idea developed in #1450 and comments of #1454 on the seeds and edge node ids.

### Checklist - did you ...

- [x] Add a [news fragment](https://landlab.readthedocs.io/en/master/development/contribution/index.html#news-entries) file entry if necessary?
- [x] Add / update tests if necessary?
- [x] Add new / update outdated documentation?
- [x] All tests have passed?
- [x] Formatted code with black?
- [x] Removed lint reported by flake8?
- [x] Sucessful documentation built? (if documentation added or modified)

### CODE REVIEW

**Comment by sebastien-lenard** [comment]
@mcflugen, So in this pull request, I only have the change of seed, right? all tests have passed, so I think i can merge it except if you have additional comments?

thank you

File: tests/grid/test_constructors.py [hunk 1]
```diff
@@ -227,8 +285,12 @@ def test_framed_voronoi_from_dict():
...
 
     assert_array_almost_equal(
-        mg.adjacent_nodes_at_node[0:2] + mg.adjacent_nodes_at_node[28:30],
-        true_adjacent_nodes_at_node_0_2 + true_adjacent_nodes_at_node_28_30,
+        np.concatenate(
+            (mg.adjacent_nodes_at_node[0:2], mg.adjacent_nodes_at_node[28:30])
+        ),
+        np.concatenate(
+            (true_adjacent_nodes_at_node_0_2, true_adjacent_nodes_at_node_28_30)
+        ),
```
Comment (mcflugen) [inline tests/grid/test_constructors.py:293]: @sebastien-lenard It seems like these tests rely on a particular seed. If so, I think they should be modified or removed.

**Comment by mcflugen** [comment]
@sebastien-lenard I made some changes to this pull request. Could you please have a look and let me know what you think?

File: landlab/graph/framed_voronoi/framed_voronoi.py [hunk 2]
```diff
@@ -67,73 +65,72 @@ def xy_of_node(
...
-        n_nodes = HorizontalRectVoronoiGraph.number_of_nodes(shape)
-        n_core_nodes = n_nodes - HorizontalRectVoronoiGraph.number_of_perimeter_nodes(
-            shape
         )
 
+        if max_move[0] < 0.0 or max_move[1] < 0.0:
+            raise ValueError("minimum spacing must be greater than node spacing")
+        if np.allclose(max_move, 0.0):
+            raise ValueError("at least one of x and y moves must be greater than zero")
+
```
Comment (mcflugen) [inline landlab/graph/framed_voronoi/framed_voronoi.py:120]: @sebastien-lenard The max moves can't be negative (i.e. the min spacing can't be greater than the initial spacing) but they can be zero in one, and only one, of the *x* or *y* directions. Does that see right to you?

File: tests/grid/test_constructors.py [hunk 3]
```diff
@@ -177,27 +177,85 @@ def test_network_from_file():
     assert mg.xy_of_reference == (12345, 678910)
 
 
+def test_framed_voronoi_edge_nodes():
```
Comment (mcflugen) [inline tests/grid/test_constructors.py:180]: I added this test to check the positions of the edge nodes, which don't depend on a random seed.

File: tests/grid/test_constructors.py [hunk 4]
```diff
@@ -177,27 +177,85 @@ def test_network_from_file():
...
+@pytest.mark.parametrize("xy_min_spacing", (1, 0, 3, (3, 4), (1, 10)))
+def test_framed_voronoi_min_spacing(xy_min_spacing):
+    grid = FramedVoronoiGrid(
+        (100, 200), xy_spacing=(3, 10), xy_min_spacing=xy_min_spacing
+    )
+
+    assert np.all(grid.length_of_link >= np.min(xy_min_spacing))
+
+
+@pytest.mark.parametrize("xy_min_spacing", (10, (3, 10), (4, 11)))
```
Comment (mcflugen) [inline tests/grid/test_constructors.py:232]: I *think* all of these min spacings should result in an error.

File: tests/grid/test_constructors.py [hunk 5]
```diff
@@ -177,27 +177,85 @@ def test_network_from_file():
...
+    assert_array_almost_equal(expected_y[:, -1], actual_y[:, -1])
+
+
+@pytest.mark.parametrize("xy_min_spacing", (1, 0, 3, (3, 4), (1, 10)))
+def test_framed_voronoi_min_spacing(xy_min_spacing):
+    grid = FramedVoronoiGrid(
+        (100, 200), xy_spacing=(3, 10), xy_min_spacing=xy_min_spacing
+    )
+
+    assert np.all(grid.length_of_link >= np.min(xy_min_spacing))
```
Comment (mcflugen) [inline tests/grid/test_constructors.py:229]: @sebastien-lenard Does this seem right to you? That is, all of the distances between connected nodes should be greater than the minimum spacing?

File: landlab/graph/framed_voronoi/framed_voronoi.py (hunk 2 above)
Comment (sebastien-lenard) [reply landlab/graph/framed_voronoi/framed_voronoi.py:120]: Yes, perfect.

File: tests/grid/test_constructors.py (hunk 5 above)
Comment (sebastien-lenard) [reply tests/grid/test_constructors.py:229]: Yes, that's right

File: tests/grid/test_constructors.py (hunk 3 above)
Comment (sebastien-lenard) [reply tests/grid/test_constructors.py:180]: That's a good idea, thank you

File: tests/grid/test_constructors.py (hunk 4 above)
Comment (sebastien-lenard) [reply tests/grid/test_constructors.py:232]: Right

File: tests/grid/test_constructors.py (hunk 1 above)
Comment (sebastien-lenard) [reply tests/grid/test_constructors.py]: Right, I removed those tests

**Comment by sebastien-lenard** [comment]
> @sebastien-lenard I made some changes to this pull request. Could you please have a look and let me know what you think?
> 
> Most of the changes are in docstrings or cleaning up some code but I've added comments to some lines where you may want to have a closer look.

Thank you a lot @mcflugen for improving this code. Everything looks ok for me and I removed the code on adjacent nodes in test_constructor.py.

**Comment by mcflugen** [comment]
Great! Thank you for looking it over!

I just want to do one more quick check of the docs, then I'll merge it.

**Comment by sebastien-lenard** [comment]
Thank you @mcflugen

**Pull request merged**